File: entity-module/customreportbuilder/ui/src/app/lib/query-compiler.ts
// Minimal client-side SQL compiler for template markers
// Supports markers: /*JOIN:<key>*/, /*WHERE:<key>*/, /*GROUP_BY*/, /*HAVING*/, /*ORDER_BY*/, /*LIMIT*/

export type JoinClause = { key: string; sql: string };      // e.g., { key: '__extra', sql: 'LEFT JOIN ...' }
export type WhereClause = { key: string; sql: string };     // e.g., { key: 'date_range', sql: "p.created_at >= DATE '2023-01-01' AND p.created_at < DATE '2026-01-01'" }

export interface QueryParts {
  joins?: JoinClause[];
  wheres?: WhereClause[];
  groupBy?: string[];  // e.g., ['p.product_id', 'c.category_name']
  having?: string[];   // e.g., ['COUNT(*) > 1']
  orderBy?: string[];  // e.g., ['p.created_at DESC']
  limit?: number | null;
}

// Basic escaping for string literals
export function sqlString(v: string): string {
  return `'${String(v).replace(/'/g, "''")}'`;
}

export function compileTemplate(baseSql: string, parts: QueryParts): string {
  let sql = baseSql;

  // Replace a marker (/*JOIN:key*/ or /*WHERE:key*/) if present
  const injectByMarker = (markerPrefix: 'JOIN' | 'WHERE', key: string, text: string) => {
    const marker = `/*${markerPrefix}:${key}*/`;
    if (sql.includes(marker)) {
      sql = sql.replace(marker, text ? `${text}\n` : '');
      return true;
    }
    return false;
  };

  // ---------- JOINS ----------
  for (const j of parts.joins ?? []) {
    if (!j?.sql) continue;
    if (!injectByMarker('JOIN', j.key, j.sql)) {
      // Fallback: insert join before the first WHERE (any casing), or at the end if none
      const m = /\bWHERE\b/i.exec(sql);
      if (m) {
        sql = `${sql.slice(0, m.index)}\n${j.sql}\n${sql.slice(m.index)}`;
      } else {
        sql = `${sql.trimEnd()}\n${j.sql}\n`;
      }
    }
  }

  // ---------- WHEREs ----------
  const groupedWheres: Record<string, string[]> = {};
  for (const w of parts.wheres ?? []) {
    if (!w?.sql) continue;
    (groupedWheres[w.key] ??= []).push(w.sql);
  }

  for (const [key, arr] of Object.entries(groupedWheres)) {
    // Combine same-key clauses under one marker
    const combined = arr.map(s => `  AND ${s}`).join('\n');

    // 1) Best: replace the matching marker spot
    if (injectByMarker('WHERE', key, combined)) continue;

    // 2) If template has "WHERE 1=1", inject right after it
    if (/\bWHERE\b\s*1\s*=\s*1/i.test(sql)) {
      sql = sql.replace(/\bWHERE\b\s*1\s*=\s*1/i, (m) => `${m}\n${combined}`);
      continue;
    }

    // 3) If template has some WHERE (without 1=1), inject right after the WHERE token
    const whereToken = /\bWHERE\b/i.exec(sql);
    if (whereToken) {
      const idx = whereToken.index + whereToken[0].length;
      sql = `${sql.slice(0, idx)}\n${combined}\n${sql.slice(idx)}`;
      continue;
    }

    // 4) No WHERE in template → create one with our clauses
    sql = `${sql.trimEnd()}\nWHERE 1=1\n${combined}\n`;
  }

  // Normalize the edge-case: "WHERE\n  AND ..." → "WHERE 1=1\n  AND ..."
  sql = sql.replace(/\bWHERE\s*\n\s*AND\b/gi, 'WHERE 1=1\n  AND');

  // ---------- GROUP BY / HAVING / ORDER BY / LIMIT ----------
  if (parts.groupBy?.length) {
    const block = `GROUP BY\n  ${parts.groupBy.join(', ')}`;
    sql = sql.replace('/*GROUP_BY*/', `${block}\n`);
  } else {
    sql = sql.replace('/*GROUP_BY*/', '');
  }

  if (parts.having?.length) {
    const block = `HAVING\n  ${parts.having.join('\n  AND ')}`;
    sql = sql.replace('/*HAVING*/', `${block}\n`);
  } else {
    sql = sql.replace('/*HAVING*/', '');
  }

  if (parts.orderBy?.length) {
    const block = `ORDER BY\n  ${parts.orderBy.join(', ')}`;
    sql = sql.replace('/*ORDER_BY*/', `${block}\n`);
  } else {
    sql = sql.replace('/*ORDER_BY*/', '');
  }

  if (typeof parts.limit === 'number' && isFinite(parts.limit)) {
    sql = sql.replace('/*LIMIT*/', `LIMIT ${parts.limit}\n`);
  } else {
    sql = sql.replace('/*LIMIT*/', '');
  }

  return sql.trim() + '\n';
}

File: entity-module/customreportbuilder/src/main/java/com/mm/customreportbuilder/reports/QueryCompiler.java
package com.mm.customreportbuilder.reports;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class QueryCompiler {
  private QueryCompiler() {}

  private static String replaceMarker(String sql, String markerPrefix, String key, String text) {
    String marker = "/*" + markerPrefix + ":" + key + "*/";
    if (sql.contains(marker)) {
      return sql.replace(marker, (text != null && !text.isBlank()) ? (text + "\n") : "");
    }
    return null; // no-op
  }

  public static String compileTemplate(String baseSql, QueryParts parts) {
    String sql = baseSql;

    // ---------- JOINS ----------
    if (parts.joins != null) {
      for (QueryParts.JoinClause j : parts.joins) {
        if (j == null || j.sql == null || j.sql.isBlank()) continue;
        String repl = replaceMarker(sql, "JOIN", j.key, j.sql);
        if (repl != null) { sql = repl; continue; }
        // Fallback: insert before the first WHERE (any casing)
        Matcher m = Pattern.compile("\\bWHERE\\b", Pattern.CASE_INSENSITIVE).matcher(sql);
        if (m.find()) {
          sql = sql.substring(0, m.start()) + "\n" + j.sql + "\n" + sql.substring(m.start());
        } else {
          sql = sql.replaceFirst("\\s*$", "") + "\n" + j.sql + "\n";
        }
      }
    }

    // ---------- WHEREs ----------
    Map<String, List<String>> grouped = new LinkedHashMap<>();
    if (parts.wheres != null) {
      for (QueryParts.WhereClause w : parts.wheres) {
        if (w == null || w.sql == null || w.sql.isBlank()) continue;
        grouped.computeIfAbsent(w.key, k -> new ArrayList<>()).add(w.sql);
      }
    }
    for (Map.Entry<String, List<String>> e : grouped.entrySet()) {
      String key = e.getKey();
      String combined = "  AND " + String.join("\n  AND ", e.getValue());

      String repl = replaceMarker(sql, "WHERE", key, combined);
      if (repl != null) { sql = repl; continue; }

      // Prefer inserting AFTER "WHERE 1=1"
      Matcher m11 = Pattern.compile("\\bWHERE\\b\\s*1\\s*=\\s*1", Pattern.CASE_INSENSITIVE).matcher(sql);
      if (m11.find()) {
        sql = new StringBuilder(sql).replace(m11.start(), m11.end(),
                sql.substring(m11.start(), m11.end()) + "\n" + combined).toString();
        continue;
      }

      // Otherwise insert right after WHERE token
      Matcher mW = Pattern.compile("\\bWHERE\\b", Pattern.CASE_INSENSITIVE).matcher(sql);
      if (mW.find()) {
        int idx = mW.end();
        sql = sql.substring(0, idx) + "\n" + combined + "\n" + sql.substring(idx);
      } else {
        // No WHERE at all → create one
        sql = sql.replaceFirst("\\s*$", "") + "\nWHERE 1=1\n" + combined + "\n";
      }
    }

    // Normalize "WHERE\nAND ..." → "WHERE 1=1\n  AND ..."
    sql = sql.replaceAll("(?i)\\bWHERE\\s*\\n\\s*AND\\b", "WHERE 1=1\n  AND");

    // ---------- GROUP BY / HAVING / ORDER BY / LIMIT ----------
    sql = injectBlock(sql, "GROUP_BY", parts.groupBy == null || parts.groupBy.isEmpty()
      ? null : "GROUP BY\n  " + String.join(", ", parts.groupBy));

    sql = injectBlock(sql, "HAVING", parts.having == null || parts.having.isEmpty()
      ? null : "HAVING\n  " + String.join("\n  AND ", parts.having));

    sql = injectBlock(sql, "ORDER_BY", parts.orderBy == null || parts.orderBy.isEmpty()
      ? null : "ORDER BY\n  " + String.join(", ", parts.orderBy));

    if (parts.limit != null && parts.limit >= 0) {
      sql = injectBlock(sql, "LIMIT", "LIMIT " + parts.limit);
    } else {
      sql = injectBlock(sql, "LIMIT", null);
    }

    // Light whitespace tidy
    sql = sql.replaceAll("[ \\t]+\\n", "\n").replaceAll("\\n{3,}", "\n\n");
    return sql.trim() + "\n";
  }

  private static String injectBlock(String sql, String key, String block) {
    String marker = "/*" + key + "*/";
    if (sql.contains(marker)) {
      return sql.replace(marker, (block != null && !block.isBlank()) ? (block + "\n") : "");
    }
    return sql; // marker absent → leave as-is
  }
}
